### DISCUSSION TITLE
Stop data access after contract agreement?

### DISCUSSION THREAD

**Discussion by ipa-big** [eclipse-edc/Connector#4662] [Q&A] [answered]
Hi,

i'm wondering how it is possible to stop giving access to an asset from consumer and provider side after a contract agreement? 

In my case I do have an asset with a data adress of type "httpdata". After the negotiation is successful and there is a contract agreement, the consumer consumes the data of the asset via the information from the related EDR. 

In case the consumer does not want access to the asset or the provider does not want to give access to the asset anymore what has to be done on provider/consumer-side to stop the access?

So far i found that I cannot delete an asset after it is linked to one contract agreement and it is not possible do delete contract agreements at all. I also tried to terminate all transfer processes that are linked to the asset but it looks like i'm still able to consume the asset, even after all transfer processes are in state "DEPROVISIONED".

Thanks for your answers

**Comment by wolf4ood** [comment]
Hi @ipa-big 

transfer termination is the right way to go. I think if you terminate transfer from the consumer mgmt api you may have stumbled upon this

https://github.com/eclipse-edc/Connector/issues/4592

**Comment by wolf4ood** [comment, marked as answer]
@ipa-big 

instead of you want contract cancellation, which means that no new transfer processes can be started on a contract, it's not supported out of the box. You need to extend EDC in order to support this feature

**Reply by ipa-big** [reply]
This is the answer that I'm looking for. 

But this means that I have to extend the data model of the contract agreement by a state (or something similar) which indicates if a agreement is still valid. Based on that state I allow or refuse the connector to initate new transfer processes. Correct?

**Reply by ndr-brt** [reply]
you can take a look on how it was implemented in Tractus-X EDC:
- [decision record](https://github.com/eclipse-tractusx/tractusx-edc/tree/main/docs/development/decision-records/2024-10-04-Retire-agreements) 
- [implementation PR](https://github.com/eclipse-tractusx/tractusx-edc/pull/1615)

**Reply by ipa-big** [reply]
good to know. 

It is available since version 0.8.0 right? I'm using 0.6.0 right now. I guess thats the reason why i did not stumble upon that feature.

**Reply by ndr-brt** [reply]
yes, Tractus-X EDC version 0.8.0 (that is built on EDC 0.10.1). 
The general advice is to stay up-to date, 0.6.0 is already quite ancient, and several security fixes were not backported
